File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
# 맨 뒤에서부터 update
# stack에 내림차순으로 넣어두고, stack맨뒤랑 numbers[i]랑 비교하기
# stack 맨뒤가 크면 answer 갱신, stack 중 numbers[i] 보다 큰게 없으면 갱신 안함(=-1)
def solution(numbers):
    answer = [-1]*len(numbers)
    stack=[numbers[-1]]
    for i in range(len(numbers)-2,-1,-1):
        if numbers[i]<stack[-1]:
            answer[i]=stack[-1]
            stack.append(numbers[i])
        else:
            while stack and numbers[i]>=stack[-1]:
                stack.pop()
            if stack:
                answer[i]=stack[-1]
            stack.append(numbers[i])
    return answer

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
// 뒤에서 부터 stack에 저장
// 1) 비어있으면 저장, 2) stack 첫번째 수가 numbers[i]보다 크면 뒤큰수! 갱신하기
// 3) stack 첫번째 수가 numbers[i]보다 작으면, stack 갱신 
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length];
        Arrays.fill(answer, -1);
        Deque<Integer> stack = new ArrayDeque<>();
        
        for (int i=numbers.length-1;i>=0;i--){
            if (stack.isEmpty()){
                stack.addFirst(numbers[i]);
                continue;
            }
            
            if (stack.peekFirst()>numbers[i]){
                answer[i]=stack.peekFirst();
                stack.addFirst(numbers[i]);
            }
            else {
                while(!stack.isEmpty() && stack.peekFirst()<=numbers[i]){
                    stack.pop();
                }
                if (!stack.isEmpty()){
                    answer[i]=stack.peekFirst();
                }
                stack.addFirst(numbers[i]);
            }
        }
        return answer;
    }
}